relay-server/src/body/peek_line.rs

```diff
@@ -0,0 +1,187 @@
+/// underlying payload and places the polled chunks back into the payload when completing.
+///
+/// If the payload does not contain a newline, the entire payload is returned by the future. To
+/// contrain this, use `limit` to set a maximum size returned by the future.
+///
+/// Resolves to `Some(Bytes)` if a newline is found within the size limit, or the entire payload is
+/// smaller than the size limit. Otherwise, resolves to `None`. Any errors on the underlying stream
+/// are returned without change.
+pub struct PeekLine {
+    payload: Payload,
+    chunks: SmallVec<[Bytes; 3]>,
+    len: usize,
+    limit: Option<usize>,
+}
+
+impl PeekLine {
+    /// Creates a new peek line future from the given payload.
+    pub fn new(payload: Payload) -> Self {
+        Self {
+            payload,
+            chunks: SmallVec::new(),
+            len: 0,
+            limit: None,
+        }
+    }
+
+    /// Adds a maximum size of the future return value.
+    ///
+    /// Note that the underlying stream may return more data than the configured limit. The future
+    /// will still never resolve more than the limit set.
+    pub fn limit(mut self, limit: usize) -> Self {
+        self.limit = Some(limit);
+        self
+    }
+
+    fn len_exceeded(&self, len: usize) -> bool {
+        if let Some(limit) = self.limit {
+            if len > limit {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    fn revert_chunks(&mut self) {
+        // unread in reverse order
+        while let Some(chunk) = self.chunks.pop() {
+            self.payload.unread_data(chunk);
+        }
+    }
+
+    fn concatenate(&self, mut len: usize) -> Bytes {
+        let mut bytes = BytesMut::with_capacity(len);
+        for chunk in &self.chunks {
+            let remaining = if len > chunk.len() { chunk.len() } else { len };
+            bytes.extend_from_slice(&chunk[..remaining]);
+            len -= remaining;
+        }
+        bytes.freeze()
+    }
+
+    fn finish(&mut self, len: usize) -> Async<Option<Bytes>> {
+        if len == 0 {
+            return Async::Ready(None);
+        }
+
+        let line = if len == 0 || self.len_exceeded(len) {
+            None
+        } else {
+            Some(self.concatenate(len))
+        };
+
+        self.revert_chunks();
+        Async::Ready(line)
+    }
+}
+
+impl futures::Future for PeekLine {
+    type Item = Option<Bytes>;
+    type Error = <SharedPayload as Stream>::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            let chunk = match self.payload.poll() {
+                Ok(Async::Ready(Some(chunk))) => chunk,
+                Ok(Async::Ready(None)) => return Ok(self.finish(self.len)),
+                result => return result,
+            };
+
+            self.chunks.push(chunk.clone());
+            if let Some(pos) = chunk.iter().position(|b| *b == b'\n') {
```

Comment:
If you're up for an adventure, I believe you could rewrite this entire type to be `PeekUntil<const b'\n'>` just because.

Reply:
Thought about that and then didn't bother. Maybe I'll come back to it one day :)